### ISSUE TITLE
Upcoming breaking change in next.js@9.5.2 ("disallow re-export all exports")

### ISSUE THREAD

**Issue by stephanschubert** [hashicorp/next-mdx-enhanced#74]
Please see https://github.com/vercel/next.js/pull/14325 for details.

**Comment by joe-bell** [comment]
Just hit this snag too, am I right in thinking it's this line? https://github.com/hashicorp/next-mdx-enhanced/blob/0ac00218efb0f0e637b7fde401b1a076cc58c5e1/loader.js#L135

**Comment by jescalan** [comment]
That definitely could be it. I'm sorry I don't think I'll have time to dig in this week, but happy to support if someone else wants to take the lead here 🙏🏼

**Comment by joe-bell** [comment]
I'd love to, but unfortunately I'm not familiar enough with how this is built to contribute 😢 

I'm not entirely sure what we want to export from this file

**Comment by NateRadebaugh** [comment]
```
Syntax error: Using `export * from '...'` in a page is disallowed. Please use `export { default } from '...'` instead.
Read more: https://err.sh/next.js/export-all-in-page
```

https://err.sh/next.js/export-all-in-page

From the link:
#### Possible Ways to Fix It

Update your page to re-export the default component only:

```ts
export { default } from './other-page'
```

If the other page uses `getServerSideProps` or `getStaticProps`, you can re-export those individually too:

```ts
export { default, getServerSideProps } from './other-page'
// or
export { default, getStaticProps } from './other-page'
// or
export { default, getStaticProps, getStaticPaths } from './other-page/[dynamic]'
```

**Comment by jescalan** [comment]
I'm honestly not entirely sure how we are going to resolve this one. The reason we use `export *` here is to allow layouts to export `getStaticProps` and/or `getServerProps` if they need to. But since this is a plugin and not an individual page as referenced in the error above, we aren't able to know for sure if a given layout does actually export either of those two, or not.

I'm curious to get @zephraph 's take on this one, if you would be so kind?

**Comment by zephraph** [comment]
Ah, yeah, that's unfortunate.

We have two options here:
1. Don't support `getStaticProps` or `getServerProps` from files with layouts. 
2. Update the babel extension to check for `getStaticProps` or `getServerProps` in the layout file and change the export to explicitly export only the things we know exist. 

The latter is likely the more correct option here, but I'm not sure what the overall performance will look like.

**Comment by jescalan** [comment]
Boo, that is what I was thinking as well, but hoping maybe a better option might come to mind. I'll look into updating the babel plugin here, since I don't want to introduce a regression if possible.

**Comment by jemjam** [comment]
I think I might have a fix for this, have been messing around with `next-mdx-enhanced` this afternoon and got this working locally. (Might just take me a few to figure out the pr.)

**Comment by jemjam** [comment]
tl;dr:

> We have two options here:
> 1. Don't support getStaticProps or getServerProps from files with layouts.
> 2. Update ... to check for getStaticProps or getServerProps ...

This afternoon I pushed up a PR (#75) that achieves 1 and is compatible with the latest version (9.5.2) of Next. Unfortunately I haven't figured out an answer to 2 yet... Open to feedback, improvements, etc. Let me know if this is helpful.

**Comment by TLadd** [comment]
Perhaps a simple solution could be adding an option to the plugin that allows passing in which layouts need to export `getStaticProps` and/or `getServerProps`? Something like

```
{
  layoutPath: "layouts",
  defaultLayout: true,
  fileExtensions: ["mdx"],
  remarkPlugins: [],
  rehypePlugins: [],
  layoutExports: {
    layout1: ["getStaticProps"],
    layout2: ["getServerProps"]
  },
  extendFrontMatter: {
    process: (mdxContent, frontMatter) => {},
    phase: "prebuild|loader|both",
  },
}
```

Obviously being able to detect it automatically would be the ideal. But this could be done quickly and get the plugin working with latest version of next.js again.

**Comment by jescalan** [comment]
I'm working on this and will have a patch out this week. Here's my current dilemma, feedback welcome.

The way to do this right would be to parse layout files using a babel plugin in order to detect whether they export either of the data fetching methods, and if they do, make sure to also export them out of the loader. I do have a local branch with this logic working. Unfortunately, this does mean adding `@babel/core` as a dependency of this library, and adding the extra parse step would be a performance regression as well, to a library that already has clear performance issues (please, think about migrating to [next-mdx-remote](https://github.com/hashicorp/next-mdx-remote) if you have not already, friends!)

If we want to support this feature in a reliable and remotely user-friendly manner, the babel dependency add is non-negotiable. However, we could avoid the performance hit by putting the "re-export layout data fetching methods" behind an option flag that would be disabled by default. Currently I feel like this is the best option we have, other than entirely cutting the feature.

I'm planning to move in this direction but still would welcome feedback!

**Comment by brianespinosa** [comment]
> please, think about migrating to [next-mdx-remote](https://github.com/hashicorp/next-mdx-remote) if you have not already, friends!

@jescalan I think there might be a couple blockers for some on moving to `next-mdx-remote`:

1. Setting up MDX files to work as routes in the `/pages` directory the same way as they do with `next-mdx-enhanced`
2. Limitation of not being able to import from an MDX file when using `next-mdx-remote`

For the first item, if the `next-mdx-remote` docs could got updated to show a suggested setup example for using an MDX file as a page route, that would lower the barrier to entry. 

For the second item, I suspect some people may have a lot of MDX files with imports. Some of those imports could be removed by declaring those components in scope with `next-mdx-remote`. I wonder if all could be resolved this way though. Hard to say without some feedback on how everyone is using `next-mdx-enhanced` as of today.

I am using both libraries in different projects. I'll probably be able to move to exclusively using `next-mdx-remote`, but I know my use case(s) does not necessarily match that of everyone.

**Comment by jescalan** [comment]
Yeah this is very fair feedback, I probably could be more clear in the docs about how to set it up the traditional way. It's definitely possible - the only real limitation is that you can't put the mdx files in `pages`, but you can just put them in another directory and they work the same, still being able to generate all the same routes as expected. And you can import from an mdx file this way, it just uses a different method.

I think the big change probably is just a little mind-bending around typical usage patterns. I will work on making this more clear in the docs though, thank you!

**Comment by brianespinosa** [comment]
> the only real limitation is that you can't put the mdx files in `pages`, but you can just put them in another directory and they work the same

Yep, this is what I am doing when this is needed. I just created an `mdxPages` directory and use them there. I think if we had some sort of "recommended" setup for that in `next-mdx-remote` that would be a big help.

> I think the big change probably is just a little mind-bending around typical usage patterns

IMO, the pattern of defining all of the component imports needed in the `components` options is actually a better developer experience. And for people on my design engineering team who are less familiar with code but still need to contribute to documentation, remembering to declare imports for components is one less thing for them to have to think about. Again... this is only one use case so it might be more problematic for people who need other dependencies in their MDX.

**Comment by johnlindquist** [comment]
@brianespinosa I would love to see the config for supporting an `mdxPages` directory that generates routes 🙏

**Comment by jescalan** [comment]
It would be super helpful if you all would open up issues in [next-mdx-remote](https://github.com/hashicorp/next-mdx-remote/issues) for these requests! 💖

**Comment by brianespinosa** [comment]
Added: https://github.com/hashicorp/next-mdx-remote/issues/27

**Comment by jescalan** [comment]
Ok, this issue is closed with tests in place and there will soon be a release out. Yay!
